### PR TITLE
Dev Tools: Source map decoder

### PR DESCRIPTION
**Pull request by barnjamin**
The VLQ encoding is a lot to ask folks to implement themselves, lets provide _some_ easier way to get the numeric pc

Lifted from @algochoi https://github.com/algochoi/teal-sourcemap-decoder

@michaeldiamant I think this would need to be implemented across the SDKs assuming we can come up with a good way to expose it.

I'm imagining with the json src_map

```py
err_pc_num = 996 # really comes from error message returned by algod
sm = SourceMap(src_map)
print(f"Teal source line number: {sm.get_line_for_pc(err_pc_num)}")
```

### CODE REVIEW

**Comment by algochoi** [comment]
Nice, I think this will work nicely when the APIs also support the disassemble endpoint [(related ticket)](https://github.com/algorand/algorand-sdk-testing/issues/170). We could add some cucumber tests in the testing repo as well to check the translation, but I'm not sure what would be the easiest way to test it (maybe output the VLQ->Int mapping in a unified JSON format?)

File: algosdk/logic.py
Comment (michaeldiamant) [inline algosdk/logic.py:294]: @barnjamin Optional:  I can imagine adding more sourcemap-related facilities later on.  For example, assuming PyTeal-to-TEAL sourcemaps exist, we may wish to provide facilities for chaining sourcemaps.

I raise the note because I'm wondering:  What do you think about relocating the functionality to a new file?

I don't feel _too_ strongly though I can imagine `logic.py` growing to be large over time.

File: run_integration.sh
Comment (barnjamin) [inline run_integration.sh:10]: Will take out before merging, want to see it pass in CI

File: algosdk/source_map.py
Comment (algochoi) [inline algosdk/source_map.py:40]: nit: adding a comment here would be helpful, e.g.

```
# Mappings may have up to 5 segments:
# Third segment represents the zero-based starting line in the original source represented.
```

File: tests/steps/other_v2_steps.py
Comment (algochoi) [inline tests/steps/other_v2_steps.py:1708]: nit: linter thing, rename `step_impl` here

**Review by algochoi — COMMENTED** [review]
Looking good - are the integration tests running? The output looks a little empty 🤔

File: algosdk/logic.py
Comment (algochoi) [inline algosdk/logic.py:5]: nit: I think this can be deleted

**Review by ahangsu — COMMENTED** [review]
just took a quick pass, and I have a minor question: do we want to point the testing branch to `sourcemap`?

File: algosdk/logic.py
Comment (michaeldiamant) [reply algosdk/logic.py:4]: Resolving - Thanks for working towards consistency across SDKs.

File: algosdk/source_map.py
Comment (michaeldiamant) [inline algosdk/source_map.py:66]: @barnjamin Minor:  Do you expect the VLQ decode + encode methods to be used _directly_ by SDK users?

My read is:  No - The methods are intended for internal API usage.  If that's the case, propose prefixing these methods with `_` to demarcate as private.
* `base64vlq_decode`
* `base64vlq_encode`

Aside:  If intended for internal usage, it might be possible to remove `base64vlq_encode`.  It's unused.

File: algosdk/source_map.py
Comment (barnjamin) [reply algosdk/source_map.py:32]: No, the line_num may be None otherwise but I'm not sure exactly when. @algochoi ?

File: algosdk/source_map.py
Comment (barnjamin) [reply algosdk/source_map.py:66]: I don't expect they'll be used outside the source map stuff, I'll make them private. 

I am a fan of deleting code but this is somewhat hairy to re-implement without a reference so in this case I'd be in favor of leaving it in unless you've got a strong opinion

File: algosdk/source_map.py
Comment (ahangsu) [reply algosdk/source_map.py:57]: ah, fair enough.

File: algosdk/source_map.py
Comment (michaeldiamant) [inline algosdk/source_map.py:14]: @barnjamin For my understanding:  What motivated exposing `delimiter` for user configuration?

From my read of the spec, the delimiter is _always_ `;`.  I'm trying to gauge if a situation exists where the user supplies a source map v3 compatible mapping with different delimiter.

Aside:  _If_ the delimiter is removed here, then I think the same change ought to happen in js-algorand-sdk.

File: algosdk/source_map.py
Comment (algochoi) [reply algosdk/source_map.py:32]: IIRC we are converting from PC->line here, so consider a mapping like: `;1;;2` (I used ints rather than VLQ for readability).

Then `pc_list` would contain `{0:None, 1: 1, 2: None, 3: 2}` (the values are decoded from `_decode_int_value`). I think I left a comment in the original decoding scheme to remind myself that doing

```
if line_num:
    blah
```

could be misleading since a `0` for line_num could be valid.

File: algosdk/source_map.py
Comment (michaeldiamant) [inline algosdk/source_map.py:19]: @barnjamin Open to discussion:  While reviewing, I realized go-algorand incidentally returns _mapping_ instead of the spec defined _mappings_ key.

I think it's most prudent to address the underlying issue (which will cause a backwards incompatibility) because the current implementation deviates from the spec.

I apologize for the inconvenience - I reviewed the PR in-question and missed the deviation (https://github.com/algorand/go-algorand/pull/3726).

Comment (algochoi) [reply algosdk/source_map.py:19]: Uhoh, good catch

Comment (barnjamin) [reply algosdk/source_map.py:19]: I can update the SDKs and tests to `mappings` but I suppose we should wait until the changes are merged into go-algorand before releasing? I can also put something in here to check for either key for now?

Comment (michaeldiamant) [reply algosdk/source_map.py]: @barnjamin Here's my thinking - What do you think?
* In go-algorand, the principled change is to:
  * Deprecate `mapping` via warning in OAS _description_, and continue populating it.  Circle back to removing it another day.
  * Add `mappings` field.  
*  In SDKs, check `mappings` first and if the key is absent, fallback to `mapping`.  When removing `mapping` server-side, we can remove the secondary look up in SDKs.

If the thinking makes sense, I can draft a story to track removal of the to-be-deprecated `mapping` field.

File: algosdk/source_map.py
Comment (michaeldiamant) [reply algosdk/source_map.py]: @barnjamin I agree with your rationale.

I am surprised there's no _well-supported_ reference implementation in Python.  Feels preferable to depend on a library here, but I didn't find one.

File: algosdk/source_map.py
Comment (michaeldiamant) [reply algosdk/source_map.py]: @barnjamin Resolving - Leaving a few notes for context.

* Thanks for making changes to support falling back between `mapping` and `mappings`.  
  * I have slight preference to prefer the non-deprecated field (`mappings`).  But, it's a minor concern and preferring `mapping` works for me.
* I opened a few stories to track deprecation:
  * Epic covering all deprecations - https://github.com/algorand/algorand-sdk-testing/issues/200
  * Story for `mapping` removal - https://github.com/algorand/algorand-sdk-testing/issues/202

File: algosdk/source_map.py
Comment (michaeldiamant) [reply algosdk/source_map.py:36]: @barnjamin Got it - I'd appreciate a double check on the spec from you + reviewers.  Provided no concerns raised, I favor removing the initialization to more closely adhere to the spec.

File: algosdk/source_map.py
Comment (michaeldiamant) [inline algosdk/source_map.py:49]: @barnjamin What's the desired behavior when the key (PC) doesn't exist?

As written, `get_line_for_pc` and `get_pcs_for_line` throw `KeyError`.  I'm double checking if the behavior is by-design.

If not, propose switching to `dict.get` and optionally, adding `Optional` return type hint to both methods.

Comment (michaeldiamant) [reply algosdk/source_map.py:49]: Aside:  I think the other SDKs warrant consideration too.  

For example, the as written Go implementation returns the Map's zero value when provided a non-existent key.  I imagine we want to return `Nil`.

File: algosdk/source_map.py
Comment (barnjamin) [reply algosdk/source_map.py]: removed

File: algosdk/source_map.py
Comment (michaeldiamant) [inline algosdk/source_map.py:34]: @barnjamin Minor:  Comment appears to be stale because the `0,0` mapping no longer exists.

```suggestion
```

File: algosdk/source_map.py
Comment (michaeldiamant) [reply algosdk/source_map.py:49]: Additionally, and optionally I'd consider adding a non-Cucumber unit test to enforce the desired behavior.

Comment (barnjamin) [reply algosdk/source_map.py:49]: I've no preference on how we handle this so I'll lean towards reviewers opinions

Comment (algochoi) [reply algosdk/source_map.py:49]: I also think we should avoid returning an `Exception`. I would propose `dict.get(pc, None)` and return a `None` or (less preferred) return a deepcopy of the dictionary so users can handle their own errors.

Comment (ahangsu) [reply algosdk/source_map.py:49]: I will up with returning `None`, but we may still want to expose `line_to_pc` or `pc_to_line` for someone wants to iterate through the structure.

File: algosdk/source_map.py
Comment (ahangsu) [inline algosdk/source_map.py:39]: nit: would be great if we could have a more elaborated comment on this part, discussing that `line_delta` would be `None` if `line_delta == 0`. if I am understanding it correctly

Comment (barnjamin) [reply algosdk/source_map.py]: I think this is correct? @algochoi?

This comment made me re-evaluate the logic I had to add pc to the line_to_pc map, I think it was a bug to only append to the map IFF the line_delta was not None

Comment (algochoi) [reply algosdk/source_map.py]: The current logic here looks correct to me 👍 

In the new encoding scheme, if the value is explicitly zero or `None`, it should equal the `last_line`.

Comment (algochoi) [reply algosdk/source_map.py]: I think you can delete the comment there: `# be careful for '0' checks!` as it might cause further confusion?

Comment (barnjamin) [reply algosdk/source_map.py]: its changed, the above snip is outdated 👍

File: algosdk/source_map.py
Comment (algochoi) [inline algosdk/source_map.py:35]:
```suggestion
            # line_delta is None if the line number has not changed
```

**Review by ahangsu — APPROVED** [review]
looking good, once Andrew's suggestion is resolved, we should probably ship it?

**Review by michaeldiamant — APPROVED** [review]
@barnjamin Leaving approval with a friendly reminder to revert branch change in `run_integration.sh` prior to merging.

**Pull request merged**
